### PR TITLE
Tweet photos

### PR DESCRIPTION
**Pull request by gfodor**
You can now send a tweet of a photo you've taken from Hubs, right within Hubs, by clicking on the "Tweet" button on photos.

image

Goes with https://github.com/mozilla/reticulum/pull/192, includes https://github.com/mozilla/hubs/pull/1455

This took way more work than I would have expected!

- Adds the DraftJS editor for tweet composition and some affordances for emojis, highlighting, and character limits.

- Adds an OAuth flow for twitter

- Fixed up a variety of issues around 2D interstitials/full screen mode being mishandled on mobile VR

- Adds a new `contentSubtype` networked field that is used to identify camera photos from normal images

- Fixes `vr_entry_type=2d_now`

- Adds a new local storage field which is an encoded list of one-time actions to perform on the next load of the room. This is used for the twitter oauth flow: after we get through oauth, we need to re-open the twitter dialog after the room loads with all the relevant info pre-filled. I made this facility fairly generic in order to future-proof it a bit since it does impact local storage.

One ugly side effect of the OAuth flow right now is it forces a page reload, so if you are trying to share a non-pinned photo your first time through the flow, it will end up getting removed from the room if you are the creator (but you can still tweet it, it turns out.)

### CODE REVIEW

**Comment by misslivirose** [comment]
Hm. The side effect of removing the image the first time seems like it will make people not want to tweet again because the mental model that they'll develop is that by tweeting, you'll remove the image. If they're logged in, could we pin it on their behalf to avoid that?

**Comment by gfodor** [comment]
We could def pin it -- I considered it but thought that might be worse since it ends up persisting the data. But all things considered perhaps its the right choice. If we cared enough I suppose I could rig up a thing where we pin it and then post-OAuth un-pin it, if it was pinned involuntarily, but that seems pretty hacky and prone to bugs.

File: src/storage/store.js [hunk 1]
```diff
@@ -105,6 +105,18 @@ export const SCHEMA = {
...
       }
+    },
+
+    onLoadActions: {
+      type: "array",
+      items: {
+        type: "object",
+        additionalProperties: false,
+        properties: {
+          action: { type: ["null", "string"] },
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/storage/store.js]: oh nice catch. previous versions of this had this as a singleton entry and so there was no way to "clear" it without breaking the schema validation other than changing the value to null. but now its a list we clear so this is no longer necessary

File: package.json [hunk 2]
```diff
@@ -53,6 +53,12 @@
     "dayjs-ext": "^2.2.0",
     "deepmerge": "^2.1.1",
     "detect-browser": "^3.0.1",
+    "draft-js": "^0.10.5",
+    "draft-js-counter-plugin": "^2.0.1",
+    "draft-js-emoji-plugin": "^2.1.1",
+    "draft-js-hashtag-plugin": "^2.0.3",
+    "draft-js-linkify-plugin": "^2.0.1",
+    "draft-js-plugins-editor": "^2.1.1",
```
[the remark this answers is not in the document]
Comment (gfodor) [reply package.json]: ah yeah I should have checked the bloat factor -- i figured it'd be non-trivial but thanks for verifying

File: src/assets/stylesheets/info-dialog.scss [hunk 3]
```diff
@@ -76,6 +76,10 @@
         border: none;
       }
     }
+
+    &__contents-overflow {
+      overflow-y: visible;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/assets/stylesheets/info-dialog.scss]: yeah the reason for this was so the emoji popup could overflow past the bottom of the dialog

File: src/react-components/tweet-dialog.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      <DialogContainer wide={true} allowOverflow={true} title="" {...this.props}>
+        <div className={styles.tweet}>
+          <div className={styles.editor} onClick={() => this.editorRef.focus()}>
+            <div className={styles.editorInner}>
+              <Editor
+                ref={r => (this.editorRef = r)}
+                editorState={this.state.editorState}
+                onChange={editorState => this.setState({ editorState })}
+                plugins={[emojiPlugin, hashtagPlugin, linkifyPlugin, counterPlugin]}
+              />
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/react-components/tweet-dialog.js]: great suggestions!

**Comment by misslivirose** [comment]
If the user is in a Discord-bound room, does the tweet have the same default text with the invitation? Is there a way that we could check that and possibly have separate default messages?

**Comment by misslivirose** [comment]
What's new copy suggestion: 

You can now share photos that you've taken from Hubs on Twitter by clicking the 'Tweet' button on photos. By default, your photos will also share an invitation link to the room you're in, so that others can easily join you, but you can also customize your message before sending your photo out into the world.

**Comment by gfodor** [comment]
Ah I am not doing anything wrt discord vs non-discord -- what would you suggest? I think it might make sense to drop the room link, but i guess its somewhat hard to know for sure. (The link will work obviously, but it'll be discord oauth gated)

**Comment by misslivirose** [comment]
I suppose I don't have data for this, but I would imagine there will be a fair amount of inaccessible room shares going on with auth-ed rooms that have links most people wouldn't be able to access rooms through. If we know a room is Discord-bound, what about if we use a generic link to the Hubs home page instead in the default copy? 

Could also add an emoji in the default message :)

**Comment by gfodor** [comment]
Yep, I think we can just change it to something simple like that. Perhaps eventually we'll want to allow bot installers to opt-in to letting us generate invite links -- so in that scenario the tweet functionality could also be a nice acquisition tool for discord server ops.

We already have hubs.mozilla.com linked in the tweet, so I guess we can just remove the second sentence (and add the duck emoji or one of your choice to all variants :))

**Comment by misslivirose** [comment]
Wow, I _completely_ skimmed over the first part of the tweet and missed that hubs.mozilla.com was already linked. 🤦‍♂️  

My vote is for the 🦆 or 🤖 emojis :)

**Comment by gfodor** [comment]
I ended up using 🐤  -- lets see how it goes :)

**Pull request merged**